In [1]:
import numpy as np
import pandas as pd
from datetime import datetime

In [2]:
UberLyft_df = pd.read_csv('cab_rides.csv')
Weather_df = pd.read_csv('weather.csv')

In [3]:
#converting the timestamp data into real date format
UberLyft_df['date_time'] = pd.to_datetime(UberLyft_df['time_stamp']/1000, unit='s')
Weather_df['date_time'] = pd.to_datetime(Weather_df['time_stamp'], unit='s')

In [4]:
#merge the datasets to refelect same time for a location, date and hour
UberLyft_df['merged_date'] = UberLyft_df.source.astype(str) +" - "+ UberLyft_df.date_time.dt.date.astype("str") +" - "+ UberLyft_df.date_time.dt.hour.astype("str")
Weather_df['merged_date'] = Weather_df.location.astype(str) +" - "+ Weather_df.date_time.dt.date.astype("str") +" - "+ Weather_df.date_time.dt.hour.astype("str")

In [5]:
Weather_df.head()

temp            location  clouds  pressure    rain  time_stamp  humidity  \
0  42.42            Back Bay     1.0   1012.14  0.1228  1545003901      0.77   
1  42.43         Beacon Hill     1.0   1012.15  0.1846  1545003901      0.76   
2  42.50   Boston University     1.0   1012.15  0.1089  1545003901      0.76   
3  42.11              Fenway     1.0   1012.13  0.0969  1545003901      0.77   
4  43.13  Financial District     1.0   1012.14  0.1786  1545003901      0.75   

    wind           date_time                           merged_date  
0  11.25 2018-12-16 23:45:01            Back Bay - 2018-12-16 - 23  
1  11.32 2018-12-16 23:45:01         Beacon Hill - 2018-12-16 - 23  
2  11.07 2018-12-16 23:45:01   Boston University - 2018-12-16 - 23  
3  11.09 2018-12-16 23:45:01              Fenway - 2018-12-16 - 23  
4  11.49 2018-12-16 23:45:01  Financial District - 2018-12-16 - 23

In [6]:
#  df_rides['date'].dt.strftime('%m').head()
#Weather_df.index = Weather_df['merged_date']

In [8]:
merged_df = pd.merge(UberLyft_df, Weather_df, on=['merged_date'])

# Join the weather date on rides data
#merged_df = UberLyft_df.join(Weather_df, on = ['merged_date'], rsuffix ='_w')

In [9]:
merged_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1265675 entries, 0 to 1265674
Data columns (total 21 columns):
distance            1265675 non-null float64
cab_type            1265675 non-null object
time_stamp_x        1265675 non-null int64
destination         1265675 non-null object
source              1265675 non-null object
price               1164996 non-null float64
surge_multiplier    1265675 non-null float64
id                  1265675 non-null object
product_id          1265675 non-null object
name                1265675 non-null object
date_time_x         1265675 non-null datetime64[ns]
merged_date         1265675 non-null object
temp                1265675 non-null float64
location            1265675 non-null object
clouds              1265675 non-null float64
pressure            1265675 non-null float64
rain                206947 non-null float64
time_stamp_y        1265675 non-null int64
humidity            1265675 non-null float64
wind                1265675 non-null f

##### As we can see the merged dataframe consists of 1265675 entries while the original dataset had 693071 entries this is because the weather data has been recorded more than one time for the same hour, and since we merge by hour (since the rides and weather do not have the exact same timestamp), then we end up with rides having different weather attributes adn therefore more the one row of ride than one. This can be seen in the following since each id can occur multiple times

In [13]:
# count of how many instances there is of each trip
merged_df['id'].value_counts()

20ddf21b-87cd-4680-a34f-611a43091b22    15
07e3f2fe-8d3f-4232-b3ab-537a8f4f0e29    15
c343160a-228f-47dc-b866-b58d116727b7    15
f13cee51-3f8b-4b60-8c15-59bcc586fdc4    15
e56702a4-7488-4a0c-8283-ea9ebfae569d    15
                                        ..
638510a4-8b7f-44f7-900e-b403a1429cea     1
eb79e419-d734-4b6f-a4a9-3650285106c3     1
05643781-6a37-4c89-aecf-7cddd7a01262     1
1f7be267-dd86-43d1-9dd5-b324a78fdc62     1
d8e31014-d15a-4f95-bc66-0ec9e4f76f6f     1
Name: id, Length: 690107, dtype: int64

In [14]:
# an example of a specific id that occurs multiple times
merged_df[merged_df['id'] == '29ce4d5f-07f9-4ce5-ada4-ae9ea49e92a6'].iloc[:,10:22]

date_time_x                 merged_date   temp  \
900564 2018-11-26 06:14:16.085999966  North End - 2018-11-26 - 6  40.54   
900565 2018-11-26 06:14:16.085999966  North End - 2018-11-26 - 6  40.55   
900566 2018-11-26 06:14:16.085999966  North End - 2018-11-26 - 6  40.61   
900567 2018-11-26 06:14:16.085999966  North End - 2018-11-26 - 6  40.53   
900568 2018-11-26 06:14:16.085999966  North End - 2018-11-26 - 6  40.55   
900569 2018-11-26 06:14:16.085999966  North End - 2018-11-26 - 6  40.55   
900570 2018-11-26 06:14:16.085999966  North End - 2018-11-26 - 6  40.53   
900571 2018-11-26 06:14:16.085999966  North End - 2018-11-26 - 6  41.16   
900572 2018-11-26 06:14:16.085999966  North End - 2018-11-26 - 6  40.64   
900573 2018-11-26 06:14:16.085999966  North End - 2018-11-26 - 6  40.54   
900574 2018-11-26 06:14:16.085999966  North End - 2018-11-26 - 6  41.16   
900575 2018-11-26 06:14:16.085999966  North End - 2018-11-26 - 6  40.58   
900576 2018-11-26 06:14:16.085999966  North End - 2018-11-26 - 6  41.16   
900577 2018-11-26 06:14:16.085999966  North End - 2018-11-26 - 6  40.57   
900578 2018-11-26 06:14:16.085999966  North End - 2018-11-26 - 6  40.83   

         location  clouds  pressure  rain  time_stamp_y  humidity  wind  \
900564  North End    1.00   1014.19   NaN    1543212884      0.91  1.25   
900565  North End    1.00   1014.19   NaN    1543213005      0.91  1.28   
900566  North End    1.00   1014.18   NaN    1543213443      0.91  1.38   
900567  North End    1.00   1014.19   NaN    1543212732      0.91  1.22   
900568  North End    1.00   1014.19   NaN    1543212945      0.91  1.27   
900569  North End    1.00   1014.19   NaN    1543212914      0.91  1.26   
900570  North End    1.00   1014.19   NaN    1543212708      0.91  1.21   
900571  North End    0.98   1014.35   NaN    1543215030      0.92  1.83   
900572  North End    1.00   1014.18   NaN    1543213742      0.92  1.45   
900573  North End    1.00   1014.19   NaN    1543212855      0.91  1.25   
900574  North End    0.98   1014.35   NaN    1543214943      0.92  1.82   
900575  North End    1.00   1014.19   NaN    1543213144      0.91  1.31   
900576  North End    0.98   1014.35   NaN    1543214984      0.92  1.82   
900577  North End    1.00   1014.19   NaN    1543212974      0.91  1.27   
900578  North End    0.92   1014.00   NaN    1543214430      0.92  1.71   

               date_time_y  
900564 2018-11-26 06:14:44  
900565 2018-11-26 06:16:45  
900566 2018-11-26 06:24:03  
900567 2018-11-26 06:12:12  
900568 2018-11-26 06:15:45  
900569 2018-11-26 06:15:14  
900570 2018-11-26 06:11:48  
900571 2018-11-26 06:50:30  
900572 2018-11-26 06:29:02  
900573 2018-11-26 06:14:15  
900574 2018-11-26 06:49:03  
900575 2018-11-26 06:19:04  
900576 2018-11-26 06:49:44  
900577 2018-11-26 06:16:14  
900578 2018-11-26 06:40:30

In [15]:
# since the weather data for each id is very similar, we choose to make a mean of all the instances by id
id_group = pd.DataFrame(merged_df.groupby('id')['temp','clouds', 'pressure', 'rain', 'humidity', 'wind'].mean())
merged_uniqueid_df = UberLyft_df.join(id_group, on = ['id'])

In [17]:
merged_uniqueid_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 693071 entries, 0 to 693070
Data columns (total 18 columns):
distance            693071 non-null float64
cab_type            693071 non-null object
time_stamp          693071 non-null int64
destination         693071 non-null object
source              693071 non-null object
price               637976 non-null float64
surge_multiplier    693071 non-null float64
id                  693071 non-null object
product_id          693071 non-null object
name                693071 non-null object
date_time           693071 non-null datetime64[ns]
merged_date         693071 non-null object
temp                690107 non-null float64
clouds              690107 non-null float64
pressure            690107 non-null float64
rain                118677 non-null float64
humidity            690107 non-null float64
wind                690107 non-null float64
dtypes: datetime64[ns](1), float64(9), int64(1), object(7)
memory usage: 95.2+ MB


In [19]:
merged_uniqueid_df['id'].value_counts()

b29737ca-eeb7-4c5f-8c1d-49e180ec05d9    1
ca4e795d-1b88-4375-9bb9-569da17557b0    1
16d49b50-7094-4a1f-b377-fe6634419c1d    1
706d0bb4-4dbb-4b06-8a1f-b1d1fe83fba0    1
3bf2fe16-ea12-4b29-a5a7-386a6a41d11a    1
                                       ..
50582b84-1a69-4bc9-a328-adec8604015a    1
db837d9d-673b-4fbe-bb31-4cbb53439730    1
63b3275e-55b5-4ef9-b91f-5736b03abd38    1
11baab35-ee4c-462c-a224-5655be64ea20    1
d6219278-3a9d-4e30-81df-38b7febc47e0    1
Name: id, Length: 693071, dtype: int64

In [20]:
merged_uniqueid_df.head()

distance cab_type     time_stamp    destination            source  price  \
0      0.44     Lyft  1544952607890  North Station  Haymarket Square    5.0   
1      0.44     Lyft  1543284023677  North Station  Haymarket Square   11.0   
2      0.44     Lyft  1543366822198  North Station  Haymarket Square    7.0   
3      0.44     Lyft  1543553582749  North Station  Haymarket Square   26.0   
4      0.44     Lyft  1543463360223  North Station  Haymarket Square    9.0   

   surge_multiplier                                    id    product_id  \
0               1.0  424553bb-7174-41ea-aeb4-fe06d4f4b9d7     lyft_line   
1               1.0  4bd23055-6827-41c6-b23b-3c491f24e74d  lyft_premier   
2               1.0  981a3613-77af-4620-a42a-0c0866077d1e          lyft   
3               1.0  c2d88af2-d278-4bfd-a8d0-29ca77cc5512   lyft_luxsuv   
4               1.0  e0126e1f-8ca9-4f2e-82b3-50505a09db9a     lyft_plus   

           name                     date_time  \
0        Shared 2018-12-16 09:30:07.890000105   
1           Lux 2018-11-27 02:00:23.677000046   
2          Lyft 2018-11-28 01:00:22.197999954   
3  Lux Black XL 2018-11-30 04:53:02.749000072   
4       Lyft XL 2018-11-29 03:49:20.223000050   

                         merged_date    temp    clouds  pressure   rain  \
0  Haymarket Square - 2018-12-16 - 9  38.460  0.290000   1022.25    NaN   
1  Haymarket Square - 2018-11-27 - 2  44.065  0.995000   1002.88  0.106   
2  Haymarket Square - 2018-11-28 - 1     NaN       NaN       NaN    NaN   
3  Haymarket Square - 2018-11-30 - 4  35.080  0.000000   1013.71    NaN   
4  Haymarket Square - 2018-11-29 - 3  37.680  0.433333    998.42    NaN   

   humidity   wind  
0  0.760000   7.68  
1  0.895000  12.63  
2       NaN    NaN  
3  0.700000   5.25  
4  0.706667  11.16

In [ ]:
# to continue see https://www.kaggle.com/cempek/uber-vs-lyft-visualizations from step 11